### PR TITLE
ENH: stats.invgauss: use Boost implementation of ppf/isf

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Supersedes gh-13665

#### What does this implement/fix?
gh-13665 implemented `ppf` and `isf` for `invgauss`, but we decided that using Boost was a better option.

This replaces our `invgauss` `ppf` and `isf` methods with those from Boost. It's faster and more accurate than main and gh-13665.

#### Additional information
I tested a62f16c2f98e5c9024d2ff7c38e1f11ccda7ee09 against main with an experiment like this.

<details>

```
import numpy as np
from scipy import stats

rng = np.random.default_rng(6424058033727397440)
qrng = stats.qmc.Halton(3, seed=rng)
sample = qrng.random(1000)
scaled = stats.qmc.scale(sample, [-8, -1, -3], [np.log10(0.5), 1, 3]) 
pz, spz, az = scaled.T
pz = 10**pz
pz[spz < 0] = 1 - pz[spz < 0]
az = 10**az
# az = np.logspace(-3, 3, 11)
# pz = np.logspace(-8, -1, 8)
# pz = np.concatenate([pz, [0.5], 1-pz[::-1]])[:, np.newaxis]

def relative_error(res, ref):
    return np.float64(abs((res-ref)/ref))

xz = stats.invgauss.ppf(pz, az)
res = stats.invgauss.cdf(xz, az)
err = relative_error(res, pz)
```

</details>

Time to evaluate `ppf` of those 1000 points:
Main: 2.08 s ± 10.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
PR: 2.18 ms ± 11.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
(similarly with isf)

Here were the error histograms:
image

That little orange bar at the right concerned me, and I found that `ppf` was very inaccurate when `x` was above `0.5` and `mu` was high (>35); the same was true of `sf`. But both were accurate when `x` was below `0.5`, though, so I added the condition that performs the evaluation in the correct tail.

Also, sometimes Boost's `ppf` and `isf` produced NaNs. In these cases, we resort to the default implementation.

After the modifications in 9bdf2513e22956ee054f1d599b9971344d4ecac7, there are no NaNs, and the maximum error is `3.200609142735203e-10`. The geometric mean of the nonzero errors is `2.4529612751351497e-15`. Ignoring the warnings produced by Boost seems safe, given the measures to replace inaccurate entries. (And yes, I think replacing is preferable to `_lazywhere` because it's so fast anyway...)

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -4117,6 +4117,26 @@ def _sf(self, x, mu):
...
         return np.exp(self._logcdf(x, mu))
 
+    def _ppf(self, x, mu):
+        with np.errstate(divide='ignore', over='ignore', invalid='ignore'):
+            x, mu = np.broadcast_arrays(x, mu)
+            ppf = _boost._invgauss_ppf(x, mu, 1)
+            i_wt = x > 0.5  # "wrong tail" - sometimes too inaccurate
+            ppf[i_wt] = _boost._invgauss_isf(1-x[i_wt], mu[i_wt], 1)
+            i_nan = np.isnan(ppf)
+            ppf[i_nan] =  super()._ppf(x[i_nan], mu[i_nan])
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4127]:
```suggestion
            ppf[i_nan] = super()._ppf(x[i_nan], mu[i_nan])
```

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -4117,6 +4117,26 @@ def _sf(self, x, mu):
...
+        return ppf
+
+    def _isf(self, x, mu):
+        with np.errstate(divide='ignore', over='ignore', invalid='ignore'):
+            x, mu = np.broadcast_arrays(x, mu)
+            isf = _boost._invgauss_isf(x, mu, 1)
+            i_wt = x > 0.5  # "wrong tail" - sometimes too inaccurate
+            isf[i_wt] = _boost._invgauss_ppf(1-x[i_wt], mu[i_wt], 1)
+            i_nan = np.isnan(isf)
+            isf[i_nan] =  super()._isf(x[i_nan], mu[i_nan])
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4137]:
```suggestion
            isf[i_nan] = super()._isf(x[i_nan], mu[i_nan])
```

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -4117,6 +4117,26 @@ def _sf(self, x, mu):
     def _cdf(self, x, mu):
         return np.exp(self._logcdf(x, mu))
 
+    def _ppf(self, x, mu):
+        with np.errstate(divide='ignore', over='ignore', invalid='ignore'):
```
Comment (mckib2) [inline scipy/stats/_continuous_distns.py:4121]: I'll have to put this on my list of un-handled Boost warnings to track down

**Comment by mckib2** [comment]
LGTM, we will see a bump in the binary size with each new Boost distribution added but hoping to solve that before the next release

File: scipy/stats/_continuous_distns.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4121]: Yeah I think they only happen when it's evaluated in the wrong tail, in which case I re-evaluate in the correct tail, so I thought they were safe to ignore for now.

**Comment by ursk** [comment]
This change produces wildly incorrect results under specific circumstances:

```
from scipy import stats
stats.invgauss(mu=2./2.001, scale=2.001).ppf(0.9643504) 
```

Expected output (what Scipy 1.9.3 produces): `6.656304078844659`
Observed output: `8.954132303260685e+248`

This broke a test in Tensorflow Probability [here](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/distributions/inverse_gaussian_test.py).


I'm not sure whether this is related to "That little orange bar at the right", or if it exposed a different edge case.

**Pull request merged**